### PR TITLE
[MTPG] Improve all_reduce and handle bwd thread support

### PR DESCRIPTION
**Pull request by kumpera**
This implements all reduce ops in all_reduce and a PG being used from a thread different than the one that created it.

We should be this >< close to getting complex training tests working.

### CODE REVIEW

File: torch/testing/_internal/distributed/multi_threaded_pg.py [hunk 1]
```diff
@@ -38,21 +40,39 @@ def ret_work(ret):
...
+
+_reduce_ops = {
+    ReduceOp.SUM: partial(binop_reduce, op=torch.sum),
+    ReduceOp.PRODUCT: partial(binop_reduce, op=torch.prod),
+    ReduceOp.MIN: partial(binop_reduce, op=torch.min),
+    ReduceOp.MAX: partial(binop_reduce, op=torch.max),
+    ReduceOp.BAND: partial(bitwise_reduce, op=torch.bitwise_and),
+    ReduceOp.BOR: partial(bitwise_reduce, op=torch.bitwise_or),
+    ReduceOp.BXOR: partial(bitwise_reduce, op=torch.bitwise_xor),
+}
```
Comment (H-Huang) [inline torch/testing/_internal/distributed/multi_threaded_pg.py:61]: nit: can these functions + dict be moved under the `AllReduce` class scope? or are they used elsewhere.

File: test/distributed/test_multi_threaded_pg.py [hunk 2]
```diff
@@ -141,9 +144,34 @@ def test_all_reduce(self):
         res_num = ((0 + self.world_size - 1) * self.world_size) / 2
         self.assertEqual(output, torch.ones(3, 3) * res_num)
 
-        # Test unimplemented error
-        with self.assertRaisesRegex(NotImplementedError, "only supports SUM on threaded pg for now"):
-            dist.all_reduce(output, op=ReduceOp.MAX)
+    def test_all_reduce_ops(self):
+        tensor = torch.tensor([dist.get_rank() + 1])
```
Comment (H-Huang) [inline test/distributed/test_multi_threaded_pg.py:148]: nit: op tests below seem like good candidates for [subtests](https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests)

**Review by H-Huang — APPROVED** [review]
Small nits, feel to merge or update at a later time

File: torch/testing/_internal/distributed/multi_threaded_pg.py (hunk 1 above)
Comment (kumpera) [reply torch/testing/_internal/distributed/multi_threaded_pg.py:61]: They will be used by ReduceScatter

**Pull request closed without merging**